deepspeed/pt/zero_optimizer_stage1.py

```diff
@@ -775,12 +876,136 @@ def state_dict(self):
+    def _partition_base_optimizer_state(self, state_key, all_partition_states):
+        partition_id = dist.get_rank(group=self.dp_process_group)
+        alignment = dist.get_world_size(group=self.dp_process_group)
+
+        flat_merged_partitions = flatten_dense_tensors_sub_partition_aligned(
+            tensor_list=all_partition_states,
+            dp=dist.get_world_size(group=self.dp_process_group),
+            max_elements_per_comm=self.max_elements_per_comm,
+            pg=self.dp_process_group)
+
+        comm_partitions, dp_sub_partitions, element_intervals, sub_partition_size, num_comm_intervals = \
+            self.get_data_parallel_sub_partitions(
+                tensor=flat_merged_partitions,
+                max_elements_per_comm=self.max_elements_per_comm,
+                world_size=dist.get_world_size(group=self.dp_process_group),
+                dp_process_group=self.dp_process_group
+            )
+
+        return [sub_partition for sub_partition in dp_sub_partitions[partition_id]]
+
+    # Compute the optimizer state partitions for the group by
+    # 1) Merging state values across the previous partitioning.
+    # 2) Repartition state values for the new partitioning
+    # 3) Return state corresponding to local partition
+    def _retrieve_group_optimizer_states(self, all_partition_states):
+        merged_optimizer_states = {}
+        for partition_state in all_partition_states:
+            for sub_partition_state in partition_state:
+                for key, value in sub_partition_state.items():
+                    if not key in merged_optimizer_states.keys():
+                        merged_optimizer_states[key] = [value]
+                    else:
+                        merged_optimizer_states[key].append(value)
+
+        group_optimizer_states = {}
+        for key, value in merged_optimizer_states.items():
+            group_optimizer_states[key] = self._partition_base_optimizer_state(
+                key,
+                value)
+
+        return group_optimizer_states
+
+    # Restore base optimizer state from checkpoint by
+    # 1) Merging optimizer state from checkpoints of all partitions
+    # 2) Extracting optimizer state for current partition from the merged state
+    # 3) Using the extracted value to directly update the base optimizer.
+    def _restore_base_optimizer_state(self, state_dict_list):
+        base_optimizer_group_states = []
+        for group_idx in range(len(self.optimizer.param_groups)):
+            all_partition_group_states = [
+                sd['base_optimizer_state'][group_idx] for sd in state_dict_list
+            ]
+            group_optimizer_states = self._retrieve_group_optimizer_states(
+                all_partition_group_states)
+            base_optimizer_group_states.append(group_optimizer_states)
+
+        for group_idx, group in enumerate(self.optimizer.param_groups):
+            for param_idx, param in enumerate(group['params']):
```

Comment:
Why do we loop over group['params'] is stage 1 and just update group['params'][0] only in stage 2?